soc/arm/st_stm32/stm32h7/power_stm32h735xx.c

```diff
@@ -0,0 +1,151 @@
+/*
+ * Copyright (c) 2021 SILA Embedded Solutions GmbH <office@embedded-solutions.at>
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+#include <zephyr.h>
+#include <pm/state.h>
+
+#include <stm32h7xx_ll_pwr.h>
+#include <stm32h7xx_ll_cortex.h>
+
+#include <logging/log.h>
+LOG_MODULE_DECLARE(soc, CONFIG_SOC_LOG_LEVEL);
+
+/* Invoke Low Power/System Off specific Tasks */
+__weak void pm_power_state_set(struct pm_state_info info)
+{
+	if (info.state != PM_STATE_SUSPEND_TO_IDLE) {
+		LOG_DBG("Unsupported power state %u", info.state);
+		return;
+	}
+
+	switch (info.substate_id) {
+	case 1: /* system mode SLEEP with VOS0*/
+		k_cpu_idle();
+		break;
+	case 2: /* system mode SLEEP with VOS1*/
+		LL_PWR_SetRegulVoltageScaling(LL_PWR_REGU_VOLTAGE_SCALE1);
+
+		while (LL_PWR_IsActiveFlag_VOS() == 0) {
+		}
+
+		k_cpu_idle();
+		break;
+	case 3: /* system mode SLEEP with VOS2*/
+		LL_PWR_SetRegulVoltageScaling(LL_PWR_REGU_VOLTAGE_SCALE2);
+
+		while (LL_PWR_IsActiveFlag_VOS() == 0) {
+		}
+
+		k_cpu_idle();
+		break;
+	case 4: /* system mode SLEEP with VOS3*/
+		LL_PWR_SetRegulVoltageScaling(LL_PWR_REGU_VOLTAGE_SCALE3);
+
+		while (LL_PWR_IsActiveFlag_VOS() == 0) {
+		}
+
+		k_cpu_idle();
+		break;
```

Comment:
Seems like these states should be sub-states of `PM_STATE_ACTIVE/PM_STATE_RUNTIME_IDLE` rather than `PM_STATE_SUSPEND_TO_IDLE`.
Cf states definitions is `include/pm/state.h`

Reply:
Moved everything to PM_STATE_RUNTIME_IDLE

Reply:
Well, the cases 5 to 10 were rather correct probably (and should be kept to PM_STATE_SUSPEND_TO_IDLE).
Cf how to handle several states and substates in `soc/arm/st_stm32/stm32u5/power.c`.

Note that, from what I know, the states are up to now used with power states
and this is the first time someone attempts to perform voltage scaling, so I can't say for sure this will work
as you expect.
Please note that another way to achieve this would be to un-correlate both pm states and VOS states,
by using separate functions.

Maybe @ceolin or @gmarull have recommendations here.

Reply:
done, so far